src/auth/flows.rs

```diff
@@ -887,3 +896,155 @@ pub async fn auth_callback(
         Err(AuthenticationError::InvalidCredentials)
     }
 }
+
+#[derive(Deserialize, Validate)]
+pub struct NewAccount {
+    #[validate(length(min = 1, max = 39), regex = "RE_URL_SAFE")]
+    pub username: String,
+    #[validate(length(min = 8, max = 256))]
+    pub password: String,
+    #[validate(email)]
+    pub email: String,
+    pub challenge: String,
+}
+
+#[post("create")]
+pub async fn create_account_with_password(
+    req: HttpRequest,
+    pool: Data<PgPool>,
+    redis: Data<deadpool_redis::Pool>,
+    new_account: web::Json<NewAccount>,
+) -> Result<HttpResponse, ApiError> {
+    new_account
+        .0
+        .validate()
+        .map_err(|err| ApiError::InvalidInput(validation_errors_to_string(err, None)))?;
+
+    if check_turnstile_captcha(&req, &new_account.challenge).await? {
+        return Err(ApiError::Turnstile);
+    }
+
+    if crate::database::models::User::get(&new_account.username, &**pool, &redis)
+        .await?
+        .is_some()
+    {
+        return Err(ApiError::InvalidInput("Username is taken!".to_string()));
+    }
+
+    let mut transaction = pool.begin().await?;
+    let user_id = crate::database::models::generate_user_id(&mut transaction).await?;
+
+    let new_account = new_account.0;
+
+    let score = zxcvbn::zxcvbn(
+        &new_account.password,
+        &[&new_account.username, &new_account.email],
+    )?;
+
+    if score.score() < 3 {
+        return Err(ApiError::InvalidInput(
+            if let Some(feedback) = score.feedback().clone().and_then(|x| x.warning()) {
+                format!("Password too weak: {}", feedback)
+            } else {
+                "Specified password is too weak! Please improve its strength.".to_string()
```

Comment:
It would be good to add a little more information here. Seems like zxcvbn doesnt give info beyond the regressed score but maybe we could add ways as examples so people dont get frustrated ("try using a longer password, or less common words" etc)

Reply:
They have a suggestions feature, but this is mainly for backend checking. On the frontend we'll have a client side view for this with suggestions.